File: backend/app/api/v1/platform.py
from typing import Any

from fastapi import APIRouter, Depends
from sqlalchemy import text
from sqlalchemy.ext.asyncio import AsyncSession

from app.db.session import get_db

router = APIRouter(prefix="/api/v1/platform", tags=["platform"])


COUNT_QUERIES = {
    "core": [
        ("organizations", "core.organizations"),
        ("users", "core.users"),
        ("documents", "core.documents"),
        ("modules", "core.app_modules"),
    ],
    "hr": [
        ("employees", "hr.employees"),
        ("teachers", "hr.teacher_profiles"),
        ("attendance_records", "hr.attendance_records"),
        ("leave_requests", "hr.leave_requests"),
    ],
    "school": [
        ("institutes", "school.institutes"),
        ("students", "school.students"),
        ("guardians", "school.guardians"),
        ("classrooms", "school.classrooms"),
        ("school_classes", "school.school_classes"),
    ],
    "timetable": [
        ("days", "school.week_days"),
        ("periods", "school.lesson_periods"),
        ("lessons", "school.timetable_slots"),
        ("versions", "school.timetable_versions"),
        ("subjects", "school.subjects"),
        ("curriculum_plans", "school.curriculum_plans"),
        ("constraints", "school.timetable_constraints"),
        ("generation_runs", "school.timetable_generation_runs"),
        ("teachers", "school.teachers"),
        ("conflicts", "school.vw_school_timetable_conflicts"),
    ],
    "education-control": [
        ("subjects", "education_control.subjects"),
        ("committees", "education_control.committees"),
        ("scores", "education_control.score_entries"),
        ("results", "education_control.results"),
    ],
    "access-mirror": [
        ("batches", "education_control.access_import_batches"),
        ("snapshots", "education_control.access_import_table_snapshots"),
    ],
    "operations": [
        ("import_jobs", "operations.import_jobs"),
        ("reports", "operations.report_templates"),
        ("feature_flags", "operations.feature_flags"),
        ("scanner_batches", "education_control.scanner_import_batches"),
    ],
}


async def table_count(db: AsyncSession, table_name: str) -> int:
    exists = await db.execute(text("SELECT to_regclass(:table_name)"), {"table_name": table_name})
    if exists.scalar_one() is None:
        return 0

    result = await db.execute(text(f"SELECT count(*) FROM {table_name}"))
    return int(result.scalar_one())


async def module_counts(db: AsyncSession, module_code: str) -> dict[str, int]:
    counts: dict[str, int] = {}
    for key, table_name in COUNT_QUERIES[module_code]:
        counts[key] = await table_count(db, table_name)
    return counts


@router.get("/modules", response_model=list[dict[str, Any]])
async def list_modules(db: AsyncSession = Depends(get_db)) -> list[dict[str, Any]]:
    modules_result = await db.execute(
        text(
            """
            SELECT code, name_ar, scope_name, route_path, status, sort_order, feature_flag
            FROM core.app_modules
            ORDER BY sort_order, name_ar
            """
        )
    )
    modules = [dict(row._mapping) for row in modules_result]

    for module in modules:
        code = module["code"]
        module["counts"] = await module_counts(db, code) if code in COUNT_QUERIES else {}
        module["acceptance"] = [
            "Migration",
            "API Contract",
            "Frontend Route",
            "Audit Ready",
        ]

    return modules


@router.get("/overview", response_model=dict[str, Any])
async def platform_overview(db: AsyncSession = Depends(get_db)) -> dict[str, Any]:
    modules = await list_modules(db)
    flags_result = await db.execute(
        text(
            """
            SELECT code, name_ar, is_enabled
            FROM operations.feature_flags
            ORDER BY code
            """
        )
    )
    flags = [dict(row._mapping) for row in flags_result]

    return {
        "product": "ERP تعليمي متكامل للمدارس والمعاهد والموظفين والكنترول",
        "version": "1.0",
        "principle": "Core Platform + Modules + Raw Mirror + Normalized Operations",
        "modules": modules,
        "feature_flags": flags,
    }

File: frontend/src/data/platform.ts
export type PlatformModule = {
  code: string;
  title: string;
  scope: string;
  description: string;
  href: string;
  status: string;
  badge: string;
  metrics: string[];
  responsibilities: string[];
};

export const platformModules: PlatformModule[] = [
  {
    code: "core",
    title: "النواة المشتركة",
    scope: "Core Platform",
    description: "المؤسسات، المستخدمون، الصلاحيات، الوثائق، التدقيق، الإعدادات، وسجل التحديثات.",
    href: "/platform",
    status: "جاهز",
    badge: "Core",
    metrics: ["organizations", "users", "documents", "modules"],
    responsibilities: ["هوية موحدة", "صلاحيات مركزية", "Audit", "Migration Registry"]
  },
  {
    code: "hr",
    title: "برنامج الموظفين",
    scope: "HR / Employee Program",
    description: "الموظفون، المدرسون، الأقسام، الوظائف، الحضور، الإجازات، المرتبات، العقود، والأداء.",
    href: "/hr",
    status: "مخطط البيانات جاهز",
    badge: "HR",
    metrics: ["employees", "teachers", "attendance_records", "leave_requests"],
    responsibilities: ["ملف موظف", "ربط المدرسين بالجدول", "حضور وانصراف", "إجازات ومرتبات"]
  },
  {
    code: "school",
    title: "برنامج المدارس",
    scope: "School Program",
    description: "المعاهد، المراحل، الصفوف، الفصول، الطلاب، أولياء الأمور، المواد، والقاعات.",
    href: "/school",
    status: "مخطط البيانات جاهز",
    badge: "School",
    metrics: ["institutes", "students", "guardians", "classrooms"],
    responsibilities: ["معاهد ومدارس", "طلاب وفصول", "أولياء أمور", "قاعات ومراحل"]
  },
  {
    code: "timetable",
    title: "الجدول المدرسي",
    scope: "School Timetable",
    description: "الأيام، الحصص، المدرسون، القاعات، الفصول، المواد، نصاب المدرس، ومنع التعارض.",
    href: "/timetable",
    status: "تشغيل فعلي",
    badge: "Timetable",
    metrics: ["days", "periods", "subjects", "curriculum_plans", "lessons", "generation_runs", "conflicts"],
    responsibilities: ["Curriculum Matrix", "Hard/Soft Constraints", "Generation Center", "Quality Score", "Approval/Publication"]
  },
  {
    code: "education-control",
    title: "برنامج الكنترول",
    scope: "Education Control",
    description: "اللجان، أرقام الجلوس، الرصد، النتائج، الدور الثاني، التفعيل، والتقارير.",
    href: "/education-control",
    status: "بدأ التنفيذ",
    badge: "Control",
    metrics: ["subjects", "committees", "scores", "results"],
    responsibilities: ["مواد ورصد", "لجان وأرقام جلوس", "نتائج وترتيب", "دور ثاني"]
  },
  {
    code: "access-mirror",
    title: "مرآة Access",
    scope: "Access Mirror",
    description: "حفظ جداول Access كاملة كمرجع خام قبل التحويل إلى جداول التشغيل المنظمة.",
    href: "/access-mirror",
    status: "بدأ التنفيذ",
    badge: "Raw Mirror",
    metrics: ["batches", "snapshots"],
    responsibilities: ["Raw Mirror", "قاموس أعمدة", "عينات صفوف", "Mapping Rules"]
  },
  {
    code: "operations",
    title: "التشغيل والتقارير",
    scope: "Operations",
    description: "السكانر، الاستيراد، الإحصائيات، الطباعة، التصدير، اللقطات، وFeature Flags.",
    href: "/operations",
    status: "مخطط البيانات جاهز",
    badge: "Ops",
    metrics: ["import_jobs", "reports", "feature_flags", "scanner_batches"],
    responsibilities: ["استيراد", "سكانر", "تقارير", "Snapshot"]
  }
];

export function findPlatformModule(code: string) {
  return platformModules.find((module) => module.code === code);
}
